### PR TITLE
Allow `os.readlink` to accept path-like objects

### PR DESCRIPTION
**Pull request by rhysparry**

### CODE REVIEW

File: stdlib/3/os/__init__.pyi [hunk 1]
```diff
@@ -319,7 +319,7 @@ def mkdir(path: _PathType, mode: int = ...) -> None: ...
 def makedirs(path: _PathType, mode: int = ...,
              exist_ok: bool = ...) -> None: ...
 def pathconf(path: _PathType, name: Union[str, int]) -> int: ...  # Unix only
-def readlink(path: AnyStr) -> AnyStr: ...
+def readlink(path: _PathType) -> AnyStr: ...
```
Comment (gvanrossum) [inline stdlib/3/os/__init__.pyi]: This is a little subtle, but `_PathType` is actually generic over `AnyStr`. And you'll need that for the return type. You can try to fix this by adding `_PathType[AnyStr]`. Hopefully that fixes the issue (which is actually caused by the inferred return type of `readlink()` being broken by to this change).

**Comment by gvanrossum** [comment]
I've got an example where I don't know what it should do.
```py
import os
class P(os.PathLike[bytes]):
    def __fspath__(self) -> bytes:
        return b'x'
print(os.readlink(P()))  # Error
```
This produces
```
__tmp__.py:5: error: Argument 1 to "readlink" has incompatible type "P"; expected "Union[str, _PathLike[str]]"
```
Without this PR it gives a different error:
```
__tmp__.py:5: error: Type argument 1 of "readlink" has incompatible value "P"
```
(At runtime it fails because the stdlib's `os.PathLike` class is not generic, but I think that's a separate issue.)

I'm not sure whether this is wort holding up this PR for though, since it's not truly a regression (more of a detour maybe :-).

**Comment by JelleZijlstra** [comment]
That seems like a mypy bug, since `P` should obviously be compatible with `PathLike[bytes]`. I'm curious where the `str`s in the error message come from, since this PR has `AnyStr`.

**Comment by gvanrossum** [comment]
Maybe, I don't have time to look into it right now. :-(

**Comment by JelleZijlstra** [comment]
Reported Guido's example above as a mypy bug in python/mypy#3644.

@rhysparry thanks for your contribution!

**Pull request merged**